R/zchunk_LA154.transportation_UCD.R

```diff
@@ -40,120 +44,390 @@ module_energy_LA154.transportation_UCD_DISABLED <- function(command, ...) {
+    IEA_fut_data_times_UCD_shares <- IEA_hist_data_times_UCD_shares %>%
+      select(-year, -value) %>%
+      distinct() %>%
+      repeat_add_columns(tibble(year = FUTURE_YEARS, value = NA))
+
+    IEA_histfut_data_times_UCD_shares <- IEA_hist_data_times_UCD_shares %>%
+      bind_rows(IEA_fut_data_times_UCD_shares) %>%
+      group_by(iso, UCD_sector, mode, size.class, UCD_technology, UCD_fuel, UCD_category, fuel) %>%
+      mutate(value = if_else(is.na(value), approx_fun(year, value, rule = 2), value))
+
+    IEA_data_times_UCD_shares_UCD_sector_agg <- IEA_histfut_data_times_UCD_shares %>%
+      group_by(iso, UCD_sector, mode, year) %>%
+      summarise(value = sum(value))
+
+    # Spreading by variable to join all at once
+    UCD_trn_data_variable_spread <- UCD_trn_data_fillout %>%
+      ungroup %>%
+      distinct() %>%
+      select(-unit) %>%
+      spread(variable, value)
+
+    ALL_ctry_var <- IEA_histfut_data_times_UCD_shares %>%
+      left_join_error_no_match(UCD_ctry, by = "iso") %>%
+      # The energy weights will be replaced by the energy weights of each mode, as many techs have 0 consumption in the base year
+      select(-value) %>%
+      fast_left_join(IEA_data_times_UCD_shares_UCD_sector_agg,
+                     by = c("iso", "UCD_sector", "mode", "year")) %>%
+      # Using a floor on the weighting factor to avoid having zero weights for any countries
+      mutate(weight_EJ = pmax(value, energy.MIN_WEIGHT_EJ)) %>%
+      # Next, match in the derived variables, specific to each individual country/sector/mode/size.class/tech/fuel, except speed
+      # There will be NA non-fuel costs which can be set to zero
+      fast_left_join(UCD_trn_data_variable_spread %>% select(-speed),
+                     by = c("UCD_sector", "mode", "size.class", "UCD_technology", "UCD_fuel", "year", "UCD_region")) %>%
+      replace_na(list(`non-fuel costs` = 0))
+
+    # Adding in speed - this is matched by the mode and (for some) size class. Match size class first
+    speed_data <- UCD_trn_data_variable_spread %>%
+      select(UCD_sector, mode, size.class, year, UCD_region, speed) %>%
+      filter(!is.na(speed))
+
+    ALL_ctry_var <- ALL_ctry_var %>%
+      fast_left_join(speed_data,
+                     by = c("UCD_sector", "mode", "size.class", "year", "UCD_region")) %>%
+      # For the missing values, join using the mode ID
+      left_join_keep_first_only(speed_data,
+                                by = c("UCD_sector", "mode", "year", "UCD_region")) %>%
+      mutate(speed.x = if_else(is.na(speed.x), speed.y, speed.x)) %>%
+      replace_na(list(speed.x = 1))
+
+    # Calculate the weighted volumes. No need to convert units because these will be aggregated and divided by weights
+    ALL_region_var <- ALL_ctry_var %>%
+      mutate(Tvkm = weight_EJ / intensity) %>%
+      mutate(Tpkm = Tvkm * `load factor`) %>%
+      mutate(Tusd = Tvkm * `non-fuel costs`) %>%
```

Comment:
They're backticks, because the variable names have spaces in them and so need to be protected.